src/System.Diagnostics.TraceSource/src/System/Diagnostics/CorrelationManager.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    public class CorrelationManager
+    {
+        internal CorrelationManager()
+        {
+        }
+
+        private readonly AsyncLocal<Stack> _idStack = new AsyncLocal<Stack>();
```

Comment:
@AlexGhiondea, try running this code:
```C#
Trace.CorrelationManager.StartLogicalOperation();
bool run = true;
Task.WaitAll(Enumerable.Range(0, 10).Select(_ => Task.Run(() =>
{
    try
    {
        while (Volatile.Read(ref run))
        {
            Trace.CorrelationManager.StartLogicalOperation(12345);
            Trace.CorrelationManager.StopLogicalOperation();
        }
    }
    finally { Volatile.Write(ref run, false); }
})).ToArray());
```
On desktop it'll run fine / forever.  With your PR I expect it fail with an exception very quickly.

> I believe we should get the same stack across async calls so that we can correlate start/stop across awaits

Desktop provides a _clone_ of the stack, which still allows you to correlate, because the stack when cloned will contain the information that was in it prior to the async operation being initiated.

> I thought you get one CallContext per thread and then the same CorrelationManager would 'stick' to the thread making multiple thread access on the same CorrelationManager not be an issue

All of the data from a call context is propagated across async points.  If there's an object in there, that then makes that same object available in the new context.  That operation can now be running concurrently with the one that kicked it off, and they both have access to the same object, so if one of them mutates the object, that mutation will be visible to the other.  That's why the Stack is cloned on the line I linked to earlier.